bindings/csharp/Regorus/Regorus.cs

```diff
@@ -0,0 +1,305 @@
+using System.Runtime.InteropServices;
+using System.Text;
+using System;
+
+#nullable enable
+namespace Regorus
+{
+	public sealed class Exception : System.Exception
+	{
+		public Exception(string? message) : base(message) { }
+	}
+
+	public sealed class Engine : System.IDisposable
+	{
+		unsafe private Regorus.Internal.RegorusEngine* E;
+		// Detect redundant Dispose() calls in a thread-safe manner.
+		// _isDisposed == 0 means Dispose(bool) has not been called yet.
+		// _isDisposed == 1 means Dispose(bool) has been already called.
+		private int isDisposed;
+
+		public Engine()
+		{
+			unsafe
+			{
+				E = Regorus.Internal.API.regorus_engine_new();
+			}
+		}
+
+		public void Dispose()
+		{
+			Dispose(disposing: true);
+
+			// This object will be cleaned up by the Dispose method.
+			// Therefore, call GC.SuppressFinalize to
+			// take this object off the finalization queue
+			// and prevent finalization code for this object
+			// from executing a second time.
+			GC.SuppressFinalize(this);
+		}
+
+		// Dispose(bool disposing) executes in two distinct scenarios.
+		// If disposing equals true, the method has been called directly
+		// or indirectly by a user's code. Managed and unmanaged resources
+		// can be disposed.
+		// If disposing equals false, the method has been called by the
+		// runtime from inside the finalizer and you should not reference
+		// other objects. Only unmanaged resources can be disposed.
+		void Dispose(bool disposing)
+		{
+			// In case _isDisposed is 0, atomically set it to 1.
+			// Enter the branch only if the original value is 0.
+			if (System.Threading.Interlocked.CompareExchange(ref isDisposed, 1, 0) == 0)
+			{
+				// If disposing equals true, dispose all managed
+				// and unmanaged resources.
+				if (disposing)
+				{
+					// No managed resource to dispose.
+				}
+
+				// Call the appropriate methods to clean up
+				// unmanaged resources here.
+				// If disposing is false,
+				// only the following code is executed.
+				unsafe
+				{
+					if (E != null)
+					{
+						Regorus.Internal.API.regorus_engine_drop(E);
+						E = null;
+					}
+				}
+			}
+		}
+
+		// Use C# finalizer syntax for finalization code.
+		// This finalizer will run only if the Dispose method
+		// does not get called.
+		~Engine() => Dispose(disposing: false);
+
+		public Engine Clone()
+		{
+			var clone = (Engine)this.MemberwiseClone();
+			unsafe
+			{
+				clone.E = Regorus.Internal.API.regorus_engine_clone(E);
+			}
+			return clone;
```

Comment:
I've not seen `MemberwiseClone` used very often. IMO a more idiomatic option to support cloning would be to define a private constructor overload:

```csharp
private unsafe Engine(Internal.RegorusEngine* engine)
{
    this.E = engine;
}

public unsafe Engine()
    : this(Regorus.Internal.API.regorus_engine_new())
{
}

public unsafe Engine Clone()
    => new(Internal.API.regorus_engine_clone(E));
```

Reply:
Nice suggestion.